### DISCUSSION TITLE
Limiting high-order discontinuous solutions with nodal DG

### DISCUSSION THREAD

**Discussion by FH-9** [mfem/mfem#4168] [Q&A]
Hi,

As far as I know, MFEM uses a purely nodal approach. I am currently writing a DG-based fluid code. Before moving to Euler equations, I would like to get everything right with 1D Burgers. I can get good solutions with resolved shocks and rarefactions comparable to the exact solutions at sufficient mesh resolutions. The problem I'm facing currently is the convergence rates that I'm getting with TVD/TVB limiters, namely, they are always worse than the ones in the papers that I'm following to build my code. Moreover, the absolute values of the errors themselves are orders of magnitudes larger than the ones obtained in those papers. And, strangely, I'm seeing even polynomial orders have way worse convergence than odd orders. My primary suspect right now is the way I implemented the limiters. I am mainly following the DG papers by Cockburn and Shu, and most of the formulations and proofs there are in terms of modal, hierarchical basis functions. After applying minmod functions on trace DOFs, their TVD/TVB limiter is applied on the P1 component of the solution, which does not really have an alternative in the nodal approach. The way I get around this problem currently is by just constructing a linear function based on the mean solution and the slope calculated using modified trace DOFs: $u^i_j = \bar{u}^i + \frac{u^{i+1/2} - u^{i-1/2}}{x^{i+1/2} - x^{i-1/2}} (x^i_j - \bar{x}^i)$. Unless I have a serious bug somewhere in my code, I believe this is what destroys the high-order accuracy.

Do you think modal approach is inherently better than the nodal one, in particular, when it comes to limiters?
How could I adjust the TVD/TVB limiters, especially the need for the P1 component, to be able to handle the nodal approach and maintain high-order accuracies? I feel like what I'm doing with that linear reconstruction is the source of my errors and low convergence rates.
Could you refer me to some papers and/or codes that use limiter procedures (not limited to TVD/TVB:) amenable to the nodal formulation employed by MFEM and maintain high-order accuracies? 
Outside the TVD/TVB area, there are are new WENO limiters for DG but it seems to me they are much more complicated and require solutions of many local optimization problems, which I'm not sure how could be implemented with MFEM.

I need some guidance on this and I am open to new ideas. Please let me know what you think.

Thank you very much.

Best,
Farhad

P.S. Not sure if this should be posted here in discussions or issues. I am posting it here but could open an issue if necessary.

**Comment by vladotomov** [comment]
See [Remhos](https://github.com/CEED/Remhos) and the papers listed in the README. Remhos uses the positive Bernstein basis and limits the DOF coefficients.

One can get optimal rates with P1 elements, there are various papers and methods that demonstrate that, for example Guermond usually uses P1 elements for his stabilizations, see Section 4.1.1 [here](https://people.tamu.edu/~guermond/PUBLICATIONS/guermond_pasquetti_JCP_2011.pdf).

**Reply by FH-9** [reply]
Thanks for your suggestions.

I always thought Bernstein basis had slow convergence in approximating a function. Isn’t that a concern with DG? How would one extract the P1 component of Bernstein? Is it as simple as getting 2 DOF coefficients? 

> One can get optimal rates with P1 elements,

Wouldn’t P1 linear elements lead to at most 2nd-order accuracy though?



Thanks.

**Reply by vladotomov** [reply]
> I always thought Bernstein basis had slow convergence in approximating a function.

The main issue with Bernstein is bad conditioning of linear systems. Otherwise the convergence should be the same as standard Lagrange elements.

> How would one extract the P1 component of Bernstein?

I'm not familiar with these limiting procedures.

> Wouldn’t P1 linear elements lead to at most 2nd-order accuracy though?

Yes. But what order are you expecting for Burgers? I'd expect at most 1.

**Reply by FH-9** [reply]
Depending on the order of approximating polynomials and the limiter, high orders can be obtained in regions excluding the shock. There are such limiters but they use Legendre polynomials and their P1 component is getting limited provided certain criteria are met. The problem is I can't get the P1 component with the nodal bases.
